### PR TITLE
Add apply configurations to generated client

### LINKED ISSUE #1811: [Feature][client] Generate apply configurations
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

The client that's generated out of the API doesn't contain apply configurations that are convenient to use with server-side apply.

### Use case

Generally, SSA is useful when multiple controllers reconcile the same resources concurrently. There are multiple use cases where this could come handy, like an external controller responsible for managing Ray cluster ingresses, or simply in e2e tests.  

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by astefanutti**
## Why are these changes needed?

This PR generates apply configurations for the Kuberay API client.

## Related issue number

Closes #1811.

## Checks

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kevin85421** [comment]
@astefanutti Do you mind giving me some keywords that I need to understand in order to have enough context to review this PR? Thanks!

**Comment by astefanutti** [comment]
@kevin85421 you'll find a very good introduction about SSA here https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/.

Here is the breakdown of the changes introduced by this PR:
* Adapt `update-codegen.sh` to generate apply configurations, note `generate-groups.sh` cannot be used, because it lacks support for external apply configurations (such as `PodTemplateSpec`), so the code-generator binaries are used directly
* Re-generate the client, with the new apply configuration packages
* Add an e2e test using SSA

File: ray-operator/test/support/core.go [hunk 1]
```diff
@@ -9,6 +9,20 @@ import (
...
 
+func Pods(t Test, namespace string, options ...Option[*metav1.ListOptions]) func(g gomega.Gomega) []corev1.Pod {
+	return func(g gomega.Gomega) []corev1.Pod {
+		listOptions := &metav1.ListOptions{}
+
+		for _, option := range options {
+			t.Expect(option.applyTo(listOptions)).To(gomega.Succeed())
+		}
+
+		jobs, err := t.Client().Core().CoreV1().Pods(namespace).List(t.Ctx(), *listOptions)
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/support/core.go]: Good catch, updated.

File: ray-operator/test/support/meta.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package support
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type ListOption interface {
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/support/meta.go]: Good catch, it's not needed. removed.

File: ray-operator/test/support/meta.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	apply(options metav1.ListOptions)
+}
+
+type labelSelector string
+
+var _ Option[*metav1.ListOptions] = (*labelSelector)(nil)
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/support/meta.go:7]: It's exactly that :)

File: ray-operator/test/e2e/support.go [hunk 4]
```diff
@@ -162,38 +133,137 @@ func rayClusterSpec() *rayv1.RayClusterSpec {
...
+							corev1.ResourceMemory: resource.MustParse("2G"),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func headPodTemplateApplyConfiguration() *corev1ac.PodTemplateSpecApplyConfiguration {
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/e2e/support.go:190]: That would be my suggestion / recommendation too. I find apply configurations fluent API style and SSA conflict management very convenient.

My original intention with the e2e tests was to use SSA, but that needed the work introduced by that PR first. We could gradually update the existing tests as you suggest.

Note the benefits of SSA are also applicable to the operator itself, but that would require more substantial work.

File: ray-operator/test/support/meta.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+// nolint: unused
+// To be removed when the false-positivity is fixed.
+func (l labelSelector) applyTo(to *metav1.ListOptions) error {
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/support/meta.go]: I think the root issue is dominikh/go-tools#1294.

There are still a number of false positive w.r.t. generics support: https://github.com/dominikh/go-tools/issues?q=is%3Aopen+label%3Afalse-positive+generic.

File: ray-operator/test/support/support.go [hunk 6]
```diff
@@ -8,9 +8,12 @@ import (
 	"github.com/onsi/gomega"
 	"github.com/onsi/gomega/format"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 var (
+	TestApplyOptions = metav1.ApplyOptions{FieldManager: "kuberay-test", Force: true}
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply ray-operator/test/support/support.go:15]: The `Force` option is used to control the resolution logic when conflicts occur. The possible strategies are documented at:

https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts

In the case of e2e tests here, the recommendation to use `Force: true` applies well. The e2e tests impersonate / emulate a real client / end-user, and what the tests specify using SSA apply configurations (e.g. RayCluster / RayJob spec) is what's owned by that client / end-user.

**Review by kevin85421 — APPROVED** [review]
Q1: I haven't reviewed the files under `pkg/client`. Are they generated files?


I want to double-check the reason why we decided to use `applyconfiguration`.
  * There are two field managers, the KubeRay operator and `kuberay-test`, in e2e tests.
  * Prior to this PR, we used `Update` without SSA, but it may have led to some 409 conflicts.
  * In this PR, we use SSA Apply with `Force: true`. The `Apply` request will always succeed.

Is my understanding correct? Thanks!

**Comment by astefanutti** [comment]
> Q1: I haven't reviewed the files under `pkg/client`. Are they generated files?

Yes, these files are generated, as the other ones that already exist in `pkg/client`.

They are "validated" by the added e2e test.

> I want to double-check the reason why we decided to use `applyconfiguration`.
> 
> * There are two field managers, the KubeRay operator and `kuberay-test`, in e2e tests.
> * Prior to this PR, we used `Update` without SSA, but it may have led to some 409 conflicts.
> * In this PR, we use SSA Apply with `Force: true`. The `Apply` request will always succeed.
> 
> Is my understanding correct? Thanks!

Yes, your understanding is correct 👍🏼.

**Comment by andrewsykim** [comment]
FYI I think CI is currently broken because the generated applyconfigurtaion was not updated after this change https://github.com/ray-project/kuberay/pull/1839

```
is missing. Running 'go mod download'.
>> Using /home/runner/go/pkg/mod/k8s.io/code-generator@v0.28.4
diffing ./hack/../pkg against freshly generated codegen
diff -Naupr ./hack/../pkg/client/applyconfiguration/ray/v1/rayservicestatus.go ./hack/../_tmp/pkg/client/applyconfiguration/ray/v1/rayservicestatus.go
--- ./hack/../pkg/client/applyconfiguration/ray/v1/rayservicestatus.go	2024-01-18 17:47:09.837[5](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:6)[6](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:7)3[7](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:8)13 +0000
+++ ./hack/../_tmp/pkg/client/applyconfiguration/ray/v1/rayservicestatus.go	2024-01-1[8](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:9) 17:46:26.866143286 +0000
@@ -6,6 +6,7 @@ package v1
 // with apply.
 type RayServiceStatusApplyConfiguration struct {
 	Applications     map[string]AppStatusApplyConfiguration `json:"applicationStatuses,omitempty"`
+	DashboardStatus  *DashboardStatusApplyConfiguration     `json:"dashboardStatus,omitempty"`
 	RayClusterName   *string                                `json:"rayClusterName,omitempty"`
 	RayClusterStatus *RayClusterStatusApplyConfiguration    `json:"rayClusterStatus,omitempty"`
 }
@@ -30,6 +31,14 @@ func (b *RayServiceStatusApplyConfigurat
 	return b
 }
 
+// WithDashboardStatus sets the DashboardStatus field in the declarative configuration to the given value
+// and returns the receiver, so that objects can be built by chaining "With" function invocations.
+// If called multiple times, the DashboardStatus field is set to the value of the last call.
+func (b *RayServiceStatusApplyConfiguration) WithDashboardStatus(value *DashboardStatusApplyConfiguration) *RayServiceStatusApplyConfiguration {
+	b.DashboardStatus = value
+	return b
+}
+
 // WithRayClusterName sets the RayClusterName field in the declarative configuration to the given value
 // and returns the receiver, so that objects can be built by chaining "With" function invocations.
 // If called multiple times, the RayClusterName field is set to the value of the last call.
diff -Naupr ./hack/../pkg/client/applyconfiguration/utils.go ./hack/../_tmp/pkg/client/applyconfiguration/utils.go
--- ./hack/../pkg/client/applyconfiguration/utils.go	2024-01-18 17:47:0[9](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:10).849563549 +0000
+++ ./hack/../_tmp/pkg/client/applyconfiguration/utils.go	2024-01-18 17:46:26.866[14](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:15)3286 +0000
@@ -[17](https://github.com/ray-project/kuberay/actions/runs/7573949808/job/20627230668?pr=1822#step:5:18),6 +17,8 @@ func ForKind(kind schema.GroupVersionKin
 		return &rayv1.AppStatusApplyConfiguration{}
 	case v1.SchemeGroupVersion.WithKind("AutoscalerOptions"):
 		return &rayv1.AutoscalerOptionsApplyConfiguration{}
+	case v1.SchemeGroupVersion.WithKind("DashboardStatus"):
+		return &rayv1.DashboardStatusApplyConfiguration{}
 	case v1.SchemeGroupVersion.WithKind("HeadGroupSpec"):
 		return &rayv1.HeadGroupSpecApplyConfiguration{}
 	case v1.SchemeGroupVersion.WithKind("HeadInfo"):
./hack/../pkg is out of date. Please run hack/update-codegen.sh
```

**Comment by andrewsykim** [comment]
https://github.com/ray-project/kuberay/pull/1847

**Comment by andrewsykim** [comment]
@astefanutti I've been finding that runing `./hack/update-codegen.sh` is really slow. Is that just for me or is it expeted?

**Comment by astefanutti** [comment]
> @astefanutti I've been finding that runing `./hack/update-codegen.sh` is really slow. Is that just for me or is it expeted?

@andrewsykim it takes around 2 minutes to complete on my setup. How long does that take on your side?

I'm not that surprised, compared to the data points I have from other projects. It may be embedding the `PodTemplateSpec` struct in multiple places contributes to increasing the generation time. It's a very large struct, which also increases the size of the CRDs significantly.

I don't see anything obvious we could do to optimise that generation time. That'd probably require to look deeper at code-generator. There is kubernetes/code-generator#69 open that could be related as well.

**Pull request merged**